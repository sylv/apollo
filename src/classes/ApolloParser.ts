import { IMDB, IMDBTitlePartial, IMDBTitleType } from "@ryanke/imdb-api";
import mem from "mem";
import { cleanFilePath } from "../helpers/cleanFilePath";
import { cleanRawTitle } from "../helpers/cleanRawTitle";
import { Logger } from "tslog";
import { properties } from "../properties/index";
import { apollo } from "../types";
import { ALL_EXTENSIONS, SUBTITLE_FILE_EXTENSIONS } from "../constants";
import { getAllMatches } from "../helpers/getAllMatches";

export class ApolloParser {
  protected matchIndexes: { start: number; end: number }[] = [];
  protected imdb = new IMDB();
  protected search = mem(this.imdb.search);
  protected log?: Logger;
  readonly options: apollo.ParserOptions;
  constructor(options: apollo.ParserOptions = {}) {
    this.options = options;
    this.log = this.options.logger;
  }

  /**
   * Parse a torrent name or complete path.
   */
  public async parse(input: string, parentData?: Partial<apollo.Parsed>): Promise<apollo.Parsed | undefined> {
    const extension = parentData ? parentData.extension : ALL_EXTENSIONS.find((ext) => input.endsWith(ext));
    const fileType = extension && SUBTITLE_FILE_EXTENSIONS.includes(extension) ? apollo.FileType.SUBTITLE : apollo.FileType.MEDIA;
    const data: Partial<apollo.Parsed> = Object.assign({}, parentData, { extension, fileType });
    const inputWithoutExt = extension && input.endsWith(extension) ? input.slice(0, -extension.length) : input;
    const cleanPath = cleanFilePath(inputWithoutExt);
    if (!cleanPath) {
      this.log?.debug(`Skipping "${input}" as it contains undesirable keywords`);
      return;
    } else {
      this.log?.debug(`Cleaned path is "${cleanPath}"`);
    }

    this.parseProperties(cleanPath, data);
    this.parseType(data);

    // with collections, the top-most title might be for the whole series.
    // trusting the file name alone is more reliable, though will mean we get less information overall.
    // this will handle something like
    // "The Hobbit & Lord of the Rings Collection/The Lord of the Rings/3-Return of the King.mkv"
    // where we would otherwise pick up "The Hobbit & Lord of the Rings Collection" as the title.
    if (parentData === undefined && data.collection) {
      // basically, if the first match in the file name is past the first
      // character in that file name, prefer the file name data as it indicates
      // the file name starts with the name of the series and not something like
      // the season number.
      const lastSep = cleanPath.lastIndexOf("/") + 1;
      const fileName = cleanPath.substring(lastSep);
      const firstFileNameMatch = this.firstMatchIndex(lastSep);
      if (firstFileNameMatch !== undefined && firstFileNameMatch > lastSep + 1) {
        const child = await this.parse(fileName, data);
        return Object.assign(data, child);
      }
    }

    data.title = this.extractTitleFromPath(cleanPath);
    if (data.title && !this.options.disableLookup) {
      const result = await this.getIMDBResult(data);
      if (result) {
        this.log?.debug(`Resolved "${data.title}" to "${result.name}" ${result.link}`);
        data.title = result.name;
        data.imdb = result;
      } else {
        this.log?.debug(`No valid results for "${data.title}"`);
      }
    }

    return data as apollo.Parsed;
  }

  /**
   * Run property parsers in the properties/ directory and add the extracted data to the provided data object.
   */
  protected parseProperties(cleanPath: string, data: Partial<apollo.Parsed>) {
    for (const property of properties) property.write(cleanPath, data, this);
    return data;
  }

  /**
   * Parse the type of title based on extracted properties.
   * data.type will be undefined if the type is unclear.
   */
  protected parseType(data: Partial<apollo.Parsed>) {
    if (data.episodeNumber?.length) data.type = IMDBTitleType.EPISODE;
    else if (data.seasons?.length || data.episodes?.length || data.seasonNumber !== undefined) data.type = IMDBTitleType.SERIES;
    else if (data.startYear && !data.endYear) data.type = IMDBTitleType.MOVIE;
    else this.log?.debug(`Cannot resolve title type based on parsed properties.`);
  }

  /**
   * Try extract the title from the cleaned file path.
   * This basically works by finding the longest string between matches,
   * so we have to match as much data as possible for accurate title extraction.
   * it will also try extact an episode name.
   */
  protected extractTitleFromPath(cleanPath: string): string | undefined {
    let previous = 0;
    const parts = [];
    const matches = this.matchIndexes.sort((a, b) => a.start - b.start);
    // we add an extra +1 because otherwise episode names won't be extracted if there isnt a match after them.
    // basically we have to do a final loop that extracts from the last match to the end of the string if
    // there isnt a match at the very end.
    for (let i = 0; i < matches.length + 1; i++) {
      const match = matches[i];
      const start = match?.start ?? cleanPath.length;
      const end = match?.end ?? cleanPath.length;
      if (previous >= start) {
        previous = end;
        continue;
      }

      const raw = cleanPath.substring(previous, start);
      const index = cleanPath.indexOf(raw);
      const title = cleanRawTitle(raw);
      // titles not near the start likely aren't relevant and are just random gibberish like
      // an episode name.
      // todo: extracting the episode name would actually be very poggers.
      if (index > 5) continue;
      if (title) parts.push({ index, title });
      previous = end;
    }

    // todo: at the moment we're just using the first one as it works pretty much every time
    // with collection handling that skips some parts of the path so we don't mismatch names.
    // it might be worth making this a bit more elaborate? maybe picking the most popular
    // result or something?
    this.log?.debug(`Extracted title candidates`, parts);
    return parts.shift()?.title;
  }

  /**
   * Get a match from a string. This automatically handles pushiung to this.matchIndexes and
   * multiple matches for you and most of the time should be preferred over string.match or RegExp.exec.
   */
  public getMatch(target: string, pattern: RegExp, returnAll: true): RegExpExecArray[];
  public getMatch(target: string, pattern: RegExp, returnAll: false): RegExpExecArray | undefined;
  public getMatch(target: string, pattern: RegExp, returnAll: boolean): RegExpExecArray | RegExpExecArray[] | undefined {
    const matches = getAllMatches(target, pattern);
    for (const match of matches) {
      const startIndex = match.index ? match.index : target.lastIndexOf(match[0]);
      if (startIndex === -1) continue;
      this.matchIndexes.push({ start: startIndex, end: startIndex + match[0].length });
    }

    if (returnAll) return matches;
    return matches[matches.length - 1];
  }

  /**
   * Get the first match index in the input.
   * For example, in "My Movie 2019", the first match would be the
   * index of the start of "2019" as that is an extracted year.
   * @param afterIndex Get the first match that is after this index.
   */
  protected firstMatchIndex(afterIndex?: number): number | undefined {
    this.matchIndexes.sort((a, b) => a.start - b.start);
    if (afterIndex === undefined) return this.matchIndexes[0]?.start;
    // checking the end here is important and intentional, otherwise
    // for example file name extraction might fail if a match crosses path parts.
    return this.matchIndexes.find((m) => afterIndex < m.end)?.start;
  }

  /**
   * Get the best search result for the given title.
   * For episodes we will look up the series title, IMDBTitleType.EPISODE is aliased to IMDBTitleType.SERIES.
   */
  protected async getIMDBResult(data: Partial<apollo.Parsed>): Promise<IMDBTitlePartial | undefined> {
    if (!data.title) throw new Error('Missing "data.title"');
    if (data.type === undefined) {
      this.log?.debug(`Cannot search without resolved title type.`);
      return;
    }

    this.log?.debug(`Searching IMDb for "${data.title}"`);
    const results = await this.search(data.title);
    const filtered: IMDBTitlePartial[] = [];
    const expectType = data.type === IMDBTitleType.EPISODE ? IMDBTitleType.SERIES : data.type;
    for (const result of results) {
      // ignore titles that don't match the expected type
      if (expectType !== result.type) continue;
      // ignore movies that don't match the extracted year (if any)
      if (data.type === IMDBTitleType.MOVIE && data.startYear && result.year && result.year !== data.startYear) continue;
      // imdb tends to return random results with extremely long names
      // when it has no clue what you're after. this avoids those results.
      // its *3 because *2 might not match "the fellowship of the ring" to "the lord of the rings: the fellowship of the ring"
      if (result.name.length > data.title.length * 3) continue;
      filtered.push(result);
    }

    if (!data.title.includes(" ")) {
      // with single-word titles, i've found IMDb struggles sometimes,
      // for example returning "Logan Lucky (2017)" before "Logan (2017)". This is a hacky way
      // to get around that.
      const haystack = filtered.slice(0, 4);
      const lowerTitle = data.title.toLowerCase();
      for (const item of haystack) {
        if (item.name.toLowerCase() === lowerTitle) return item;
      }
    }

    return filtered.shift();
  }
}
