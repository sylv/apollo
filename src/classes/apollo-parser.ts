import { search, SearchResult, TitleType } from "@ryanke/imdb-api";
import mem from "mem";
import { IMAGE_EXTENSIONS, SUBTITLE_EXTENSIONS, VIDEO_EXTENSIONS } from "../constants";
import { cleanFilePath } from "../helpers/clean-file-path";
import { cleanRawTitle } from "../helpers/clean-raw-title";
import { getAllMatches } from "../helpers/get-all-matches";
import { properties } from "../properties";
import { ApolloLogger, ApolloOutput, ApolloParserOptions, FileType } from "../types";

export interface ApolloMatch {
  start: number;
  end: number;
  countForChildFiltering: boolean;
}

export class ApolloParser {
  protected search = mem(search);
  protected log?: ApolloLogger;
  readonly matchIndexes: ApolloMatch[] = [];
  readonly options: ApolloParserOptions;
  constructor(options: ApolloParserOptions = {}) {
    this.options = options;
    this.log = this.options.logger;
  }

  /**
   * Parse a torrent name or complete path.
   * @param parentData internal
   */
  public async parse(input: string, parentData?: Partial<ApolloOutput>): Promise<ApolloOutput | undefined> {
    const extractedExt = this.getExtensionAndFileType(input);
    const extension = parentData?.extension ?? extractedExt?.extension;
    const fileType = extractedExt?.fileType;
    const data: Partial<ApolloOutput> = Object.assign({}, parentData, { extension, fileType });
    const inputWithoutExt = extension && input.endsWith(extension) ? input.slice(0, -extension.length) : input;
    const cleanPath = cleanFilePath(inputWithoutExt);
    if (!cleanPath) {
      this.log?.debug(`Skipping "${input}" as it contains undesirable keywords`);
      return;
    } else {
      this.log?.debug(`Cleaned path is "${cleanPath}"`);
    }

    this.parseProperties(cleanPath, data);
    this.parseTitleType(data);

    // with collections, the top-most title might be for the whole series.
    // trusting the file name alone is more reliable, though will mean we get less information overall.
    // this will handle something like
    // "The Hobbit & Lord of the Rings Collection/The Lord of the Rings/3-Return of the King.mkv"
    // where we would otherwise pick up "The Hobbit & Lord of the Rings Collection" as the title.
    if (parentData === undefined && data.collection) {
      // basically, if the first match in the file name is past the first
      // character in that file name, prefer the file name data as it indicates
      // the file name starts with the name of the series and not something like
      // the season number.
      const lastSep = cleanPath.lastIndexOf("/") + 1;
      const fileName = cleanPath.substring(lastSep);
      const firstFileNameMatch = this.firstMatchIndex(lastSep);
      if (firstFileNameMatch !== undefined && firstFileNameMatch > lastSep + 1) {
        const child = await this.parse(fileName, data);
        return Object.assign(data, child);
      }
    }

    data.title = this.extractTitleFromPath(cleanPath, data.titleType);
    if (data.title && !this.options.disableLookup) {
      const result = await this.getIMDBResult(data);
      if (result) {
        this.log?.debug(`Resolved "${data.title}" to "${result.name}" https://imdb.com/title/${result.id}`);
        data.title = result.name;
        data.imdb = result;
      } else {
        this.log?.debug(`No valid results for "${data.title}"`);
      }
    }

    return data as ApolloOutput;
  }

  /**
   * Run property parsers in the properties/ directory and add the extracted data to the provided data object.
   */
  protected parseProperties(cleanPath: string, data: Partial<ApolloOutput>) {
    for (const property of properties) property.write(cleanPath, data, this);
    return data;
  }

  /**
   * Parse the type of title based on extracted properties.
   * data.type will be undefined if the type is unclear.
   */
  protected parseTitleType(data: Partial<ApolloOutput>) {
    if (data.episodeNumber?.length) data.titleType = TitleType.EPISODE;
    else if (data.seasons?.length || data.episodes?.length || data.seasonNumber !== undefined) data.titleType = TitleType.SERIES;
    else if (data.startYear && !data.endYear) data.titleType = TitleType.MOVIE;
    else {
      this.log?.debug(`Cannot resolve title type based on parsed properties.`);
    }

    if (data.titleType !== undefined && !data.fileType) {
      data.fileType = FileType.Video;
    }
  }

  /**
   * Try extract the title from the cleaned file path.
   * This basically works by finding the longest string between matches,
   * so we have to match as much data as possible for accurate title extraction.
   * it will also try extract an episode name.
   */
  protected extractTitleFromPath(cleanPath: string, titleType?: TitleType): string | undefined {
    const titleCandidates: string[] = [];
    this.matchIndexes.sort((a, b) => a.start - b.start);
    for (let matchIndex = 0; matchIndex < this.matchIndexes.length; matchIndex++) {
      const match = this.matchIndexes[matchIndex];
      const previousMatch = this.matchIndexes[matchIndex - 1];
      const between = cleanPath.substring(previousMatch?.end ?? 0, match.start);
      if (!between.trim()) {
        continue;
      }

      const clean = cleanRawTitle(between);
      if (clean) {
        titleCandidates.push(clean);
      }
    }

    this.log?.debug(`Extracted title candidates`, titleCandidates);
    if (!titleCandidates[0]) {
      if (!this.matchIndexes[0]) return cleanRawTitle(cleanPath);
      return;
    }

    if (titleType === TitleType.MOVIE) {
      // for movies, there is no episode name that might get confused for a title,
      // so returning the furthest-right match is probably gonna be fine.
      const longest = titleCandidates.reduce((a, b) => (a.length > b.length ? a : b));
      const last = titleCandidates[titleCandidates.length - 1];
      if (longest.length >= last.length * 2) return longest;
      return last;
    }

    return titleCandidates[0];
  }

  /**
   * Get a match from a string. This automatically handles pushing to this.matchIndexes and
   * multiple matches for you and most of the time should be preferred over string.match or RegExp.exec.
   */
  public getMatch(target: string, pattern: RegExp, returnAll: true, countForChildFiltering?: boolean): RegExpExecArray[];
  public getMatch(target: string, pattern: RegExp, returnAll: false, countForChildFiltering?: boolean): RegExpExecArray | undefined;
  public getMatch(target: string, pattern: RegExp, returnAll: boolean, countForChildFiltering = true): RegExpExecArray | RegExpExecArray[] | undefined {
    const matches = getAllMatches(target, pattern);
    for (const match of matches) {
      const startIndex = match.index ? match.index : target.lastIndexOf(match[0]);
      if (startIndex === -1) continue;
      this.matchIndexes.push({ start: startIndex, end: startIndex + match[0].length, countForChildFiltering });
    }

    if (returnAll) return matches;
    return matches[matches.length - 1];
  }

  /**
   * Get the first match index in the input.
   * For example, in "My Movie 2019", the first match would be the
   * index of the start of "2019" as that is an extracted year.
   * @param afterIndex Get the first match that is after this index.
   */
  protected firstMatchIndex(afterIndex?: number): number | undefined {
    this.matchIndexes.sort((a, b) => a.start - b.start);
    if (afterIndex === undefined) return this.matchIndexes[0]?.start;
    // checking the end here is important and intentional, otherwise
    // for example file name extraction might fail if a match crosses path parts.
    return this.matchIndexes.find((m) => m.countForChildFiltering && afterIndex < m.end)?.start;
  }

  /**
   * Get the best search result for the given title.
   * For episodes we will look up the series title, TitleType.EPISODE is aliased to TitleType.SERIES.
   */
  protected async getIMDBResult(data: Partial<ApolloOutput>): Promise<SearchResult | undefined> {
    if (!data.title) throw new Error('Missing "data.title"');
    if (data.titleType === undefined) {
      this.log?.debug(`Cannot search without resolved title type.`);
      return;
    }

    this.log?.debug(`Searching IMDb for "${data.title}"`);
    const results = await this.search(data.title);
    const filtered: SearchResult[] = [];
    const expectType = data.titleType === TitleType.EPISODE ? TitleType.SERIES : data.titleType;
    for (const result of results) {
      // ignore titles that don't match the expected type
      if (expectType !== result.type) continue;
      // ignore movies that don't match the extracted year (if any)
      if (data.titleType === TitleType.MOVIE && data.startYear && result.year && result.year !== data.startYear) continue;
      // imdb tends to return random results with extremely long names
      // when it has no clue what you're after. this avoids those results.
      // its *3 because *2 might not match "the fellowship of the ring" to "the lord of the rings: the fellowship of the ring"
      if (result.name.length > data.title.length * 3) continue;
      filtered.push(result);
    }

    if (!data.title.includes(" ")) {
      // with single-word titles, i've found IMDb struggles sometimes,
      // for example returning "Logan Lucky (2017)" before "Logan (2017)". This is a hacky way
      // to get around that.
      const haystack = filtered.slice(0, 4);
      const lowerTitle = data.title.toLowerCase();
      for (const item of haystack) {
        if (item.name.toLowerCase() === lowerTitle) return item;
      }
    }

    return filtered.shift();
  }

  private getExtensionAndFileType(input: string): { extension: string; fileType: FileType } | undefined {
    const haystack = input.toLowerCase();
    const videoExtension = VIDEO_EXTENSIONS.find((ext) => haystack.endsWith(ext));
    if (videoExtension) {
      return { extension: videoExtension, fileType: FileType.Video };
    }

    const imageExtension = IMAGE_EXTENSIONS.find((ext) => haystack.endsWith(ext));
    if (imageExtension) {
      return { extension: imageExtension, fileType: FileType.Image };
    }

    const subtitleExtension = SUBTITLE_EXTENSIONS.find((ext) => haystack.endsWith(ext));
    if (subtitleExtension) {
      return { extension: subtitleExtension, fileType: FileType.Subtitle };
    }
  }
}
