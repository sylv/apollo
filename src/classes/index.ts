export * from "./Apollo";
export * from "./ApolloParser";
