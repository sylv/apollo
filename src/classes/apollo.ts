import fs from "fs-extra";
import path from "path";
import rrdir from "rrdir";
import sanitize from "sanitize-filename";
import { SUBTITLE_EXTENSIONS } from "../constants";
import { ApolloLogger, ApolloOptions, ApolloOutput } from "../types";
import { ApolloParser } from "./apollo-parser";

export class Apollo {
  protected readonly options: ApolloOptions;
  protected readonly log?: ApolloLogger;
  protected createdDirectories = new Set();
  protected handledFiles = new Set();

  constructor(options: ApolloOptions) {
    this.options = options;
    this.log = options.logger;
    if (this.options.disableLookup) {
      this.log?.info(`IMDb queries are disabled due to --disable-lookup.`);
    }
  }

  /**
   * Run the parser with the given options.
   */
  async run() {
    let checkedCount = 0;
    let newCount = 0;

    for await (const file of rrdir(this.options.input, { stats: true })) {
      if (file.directory) continue;
      const isSubtitle = SUBTITLE_EXTENSIONS.some((ext) => file.path.endsWith(ext));
      if (!isSubtitle && file.stats!.size < this.options.minSize) {
        this.log?.debug(`Skipping "${file.path}" as it is too small and not a subtitle file`);
        continue;
      }

      const inputPath = file.path.slice(this.options.input.length + 1);
      const parser = new ApolloParser(this.options);
      const parsed = await parser.parse(inputPath);
      if (!parsed || !parsed.title) {
        this.log?.warn(`Skipping "${file.path}" as no data could be extracted. Run with --debug for more info.`);
        continue;
      }

      const output = this.getFileOutputPath(parsed);
      // can't create a path that already exists
      if (this.handledFiles.has(output.path)) {
        this.log?.warn(`Skipping "${file.path}" as the output path "${output.path}" already exists`);
        continue;
      }

      if (!this.createdDirectories.has(output.directory)) {
        await fs.mkdirp(output.directory);
        this.createdDirectories.add(output.directory);
      }

      const action = this.options.move ? "moving" : "linking";
      try {
        checkedCount += 1;
        if (!this.options.dryRun) {
          // move or create symlinks based on options
          if (this.options.move) await fs.move(file.path, output.path);
          else fs.symlink(file.path, output.path);
        }

        this.log?.info(`${action} "${file.path}" -> "${output.path}"`);
        this.handledFiles.add(output.path);
        newCount += 1;
      } catch (error: any) {
        if (error.code === "EPERM") {
          this.log?.error(`EPERM: permission error ${action} "${file.path}" -> "${output.path}".`);
          return process.exit(1);
        } else if (error.code === "EEXIST") {
          this.log?.debug(`Failed ${action} "${output.path}" as the destination already exists.`);
          continue;
        }

        throw error;
      }
    }

    this.log?.info(`Checked ${checkedCount.toLocaleString()} files and moved ${newCount.toLocaleString()} new files`);
  }

  /**
   * Get the path a parsed title should be outputted to.
   */
  protected getFileOutputPath(parsed: ApolloOutput) {
    if (!parsed.title) throw new Error(`Missing title on input`);
    const title = sanitize(parsed.title);
    let outputPath: string;

    if (parsed.episodeNumber?.length) {
      const parentDir = parsed.seasonNumber ? `Season ${parsed.seasonNumber}` : "Unknown Season";
      const paddedSeason = parsed.seasonNumber ? "S" + parsed.seasonNumber.toString().padStart(2, "0") : "";
      const paddedEpisode = parsed.episodeNumber.map((n) => n.toString().padStart(2, "0")).join("-");
      const loc = `TV Shows/${title}/${parentDir}/${title} ${paddedSeason}E${paddedEpisode}${parsed.extension}`;
      outputPath = path.join(this.options.output, loc);
    } else {
      const yearSuffix = parsed.startYear && ` (${parsed.startYear})`;
      const joined = sanitize(yearSuffix ? title + yearSuffix : title);
      outputPath = path.join(this.options.output, `Movies/${joined}/${joined}${parsed.extension}`);
    }

    return {
      directory: path.dirname(outputPath),
      path: outputPath,
    };
  }
}
