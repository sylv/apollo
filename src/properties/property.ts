import { ApolloParser } from "../classes/apollo-parser";
import { ApolloOutput } from "../types";

export abstract class Property<Key extends keyof ApolloOutput> {
  readonly key?: Key;

  extract?(cleanPath: string, parser: ApolloParser): ApolloOutput[Key] | undefined;
  write(cleanPath: string, data: Partial<ApolloOutput>, parser: ApolloParser): Partial<ApolloOutput> {
    if (!this.key) throw new Error(`Missing "key" on ${this.constructor.name}`);
    if (!this.extract) throw new Error(`Missing "extract" on ${this.constructor.name} with default "write"`);
    const output = this.extract(cleanPath, parser);
    if (output !== undefined) {
      // previously this assigned it regardless but if we set undefined values to undefined,
      // Object.assign will overwrite set values with undefined. This is bad when we handle collections,
      // so not defining it if its undefined might allow more data in.
      (data as any)[this.key] = output;
    }

    return data;
  }
}
