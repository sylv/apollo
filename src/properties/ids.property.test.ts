import { ApolloParser } from "../classes/apollo-parser";
import { IdType, PropertyIds } from "./ids.property";

let parser: ApolloParser;

beforeEach(() => (parser = new ApolloParser()));

it("should extract ids", () => {
  const extractor = new PropertyIds();
  expect(extractor.extract("[id=t3_vavr7h]", parser)).toMatchSnapshot();
  expect(extractor.extract("1525508539468419074_1", parser)).toMatchSnapshot();
  expect(extractor.extract("1525508539468419074_1", parser)).toMatchSnapshot();
  expect(extractor.extract("test-ph5d48409d154c4", parser)).toMatchSnapshot();
  expect(extractor.extract("https://www.pornhub.com/view_video.php?viewkey=ph5d48409d154c4", parser)).toMatchSnapshot();
});
