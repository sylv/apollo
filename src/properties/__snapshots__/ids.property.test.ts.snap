// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should extract ids 1`] = `
Array [
  Object {
    "id": "t3_vavr7h",
    "type": "REDDIT",
    "url": "https://www.reddit.com/comments/t3_vavr7h",
  },
]
`;

exports[`should extract ids 2`] = `
Array [
  Object {
    "id": "1525508539468419074",
    "type": "TWITTER",
    "url": "https://twitter.com/i/status/1525508539468419074",
  },
]
`;

exports[`should extract ids 3`] = `
Array [
  Object {
    "id": "1525508539468419074",
    "type": "TWITTER",
    "url": "https://twitter.com/i/status/1525508539468419074",
  },
]
`;

exports[`should extract ids 4`] = `
Array [
  Object {
    "id": "ph5d48409d154c4",
    "type": "PORNHUB",
    "url": "https://www.pornhub.com/view_video.php?viewkey=ph5d48409d154c4",
  },
]
`;

exports[`should extract ids 5`] = `
Array [
  Object {
    "id": "ph5d48409d154c4",
    "type": "PORNHUB",
    "url": "https://www.pornhub.com/view_video.php?viewkey=ph5d48409d154c4",
  },
]
`;
